Lib/logging/__init__.py

```diff
@@ -1555,36 +1555,41 @@ def findCaller(self, stack_info=False, stacklevel=1):
         Find the stack frame of the caller so that we can note the source
         file name, line number and function name.
         """
-        f = currentframe()
+        f = _currentframe()
         #On some versions of IronPython, currentframe() returns None if
         #IronPython isn't run with -X:Frames.
-        if f is not None:
-            f = f.f_back
-        orig_f = f
-        while f and stacklevel > 1:
-            f = f.f_back
-            stacklevel -= 1
-        if not f:
-            f = orig_f
-        rv = "(unknown file)", 0, "(unknown function)", None
-        while hasattr(f, "f_code"):
+        if f is None:
+            return "(unknown file)", 0, "(unknown function)", None
+        while stacklevel > 0:
+            next_f = f.f_back
+            if next_f is None:
+                ##TODO: We've got options here
+                ## If we want to use the last (deepest) frame:
+                break
+                ## If we want to mimic the warnings module:
+                #return ("sys", 1, "(unknown function)", None)
+                ## If we want to be pedantic:
+                #raise ValueError("call stack is not deep enough")
+            f = next_f
             co = f.f_code
             filename = os.path.normcase(co.co_filename)
-            if filename == _srcfile:
-                f = f.f_back
+            # The following condition is partially copied from
+            # warnings._is_internal_frame
+            if filename == _srcfile or (
+                "importlib" in filename and "_bootstrap" in filename):
+            ):
```

Comment:
Why start skipping `importlib` and `_bootstrap`? Does the warning module have a reason for that? Why would the logging module benefit from it? If so, I would it would be better just call their function `_is_internal_frame` instead of copying its contents.

Reply:
Initially, the `warnings` module was used for it, but then I realized that no underscore methods of other modules are used in this file. Therefore, this implementation of `logging` could in principle be used in Python implementations that have a different implementation of `warnings`. For this reason, I chose to duplicate that piece of code. I noted that the `currentframe` code was initially [lifted from `inspect`](https://github.com/python/cpython/commit/829dc51a050a68ca79cdde298697d0344639bd14).

Reply:
Skipping `importlib._bootstrap` [was done](https://bugs.python.org/issue24305) to allow module-level warnings (here: log messages) with stacklevels that point to the importing module and not to the import mechanism. A useful thing, I'd say.

Reply:
Thanks for explaining it. I am not sure if it would be a problem for the logging module to show the stack trace from `importlib` or `_bootstrap` as it was a problem for the warning module. What if I am using the logging module inside `importlib`? Disregarding this last scenario, I would wait to see what someone more experienced with Python library think about this.

Just out of curiosity, did you have trouble with the logging module showing stack traces of `importlib` or `_bootstrap` which lead you to bring this checking from the warning module or you just happened to know about their protection and thought it would be good to add it here too?

Reply:
It is not "`importlib` _or_ `_bootstrap`" that is checked for, but "`importlib` _and_ `_bootstrap`" (i.e. `importlib._bootstrap`). There is no logging in `Lib/importlib/_bootstrap.py` and that module doesn't import anything (for obvious reasons), so I think we can safely say that the logging module is not used in `importlib._bootstrap` :-).

I ran into the stacklevel oddity mentioned in [bpo-45171](https://bugs.python.org/issue45171) and read that `stacklevel` was supposed to work like "the equivalent [parameter] in the warnings module". Thus I read how `warnings` chases the stack and copied the logic to `logging`.